eth/dao_hard_fork.go

```diff
@@ -0,0 +1,197 @@
+package eth
+
+// WARNING: THIS CODE IS WORK IN PROGRESS AND NEEDS EXTENSIVE REVIEW
+//
+// Utility functions for preparing and enacting a hard fork to revert the hack of "The DAO"
+// according to proposed spec at:
+// https://docs.google.com/document/d/1VfuAH7Zf0UQmuVw1o7cTPbNtp1wFYzlN1tBKV6SbVSI
+//
+
+import (
+	"encoding/json"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/internal/dao"
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+)
+
+var (
+	theDAOAddr = common.HexToAddress("bb9bc244d798123fde783fcc1c72d3bb8c189413")
+	// TODO: set these after community consensus
+	theDAOReplacementCode = []byte{0x00}
+	hardForkBlock         = big.NewInt(999999999)
+)
+
+type DAOHardForkData struct {
+	ChildDAOs             []common.Address
+	ExtraBalances         []common.Address
+	TheDAOReplacementCode []byte
+}
+
+func TestRunHardFork(eth *Ethereum) error {
+	hfData, err := prepareHardForkData(eth)
+	if err != nil {
+		return err
+	}
+
+	// log JSON representation of hard fork data for manual inspection
+	logHardForkData(hfData, eth)
+
+	// uncomment to (for testing) apply hard fork on current state:
+	// err = applyHardFork(hfData, eth)
+	return err
+}
+
+func logHardForkData(hfData *DAOHardForkData, eth *Ethereum) {
+	j, _ := json.Marshal(hfData)
+	glog.V(logger.Info).Infof("Generated hard fork data: %s", j)
+
+	statedb, err := eth.BlockChain().State()
+	if err != nil {
+		return
+	}
+
+	// only accumulated for logging purposes
+	total := big.NewInt(0)
+	for i := 0; i < len(hfData.ChildDAOs); i++ {
+		acc := statedb.GetAccount(hfData.ChildDAOs[i])
+		total.Add(total, acc.Balance())
+	}
+
+	for i := 0; i < len(hfData.ExtraBalances); i++ {
+		acc := statedb.GetAccount(hfData.ExtraBalances[i])
+		total.Add(total, acc.Balance())
+	}
+
+	glog.V(logger.Info).Infof("Total wei in all accounts: %v", total)
+}
+
+func prepareHardForkData(eth *Ethereum) (*DAOHardForkData, error) {
+	hfData := &DAOHardForkData{
+		TheDAOReplacementCode: theDAOReplacementCode,
+	}
+
+	theDAO, err := dao.NewDAO(theDAOAddr, NewContractBackend(eth))
+	if err != nil {
+		glog.V(logger.Info).Infof("could not instantiate DAO: %v", err)
+		return nil, err
+	}
+
+	// create list of "The DAO" and all its children, recursively
+	daos, err := traverseDAOs(theDAO, eth)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(logger.Info).Infof("DAOs traversed: %v", len(daos))
+
+	// and their extraBalance accounts
+	for i := 0; i < len(daos); i++ {
+		bal, err := daos[i].ExtraBalance(nil)
+		if err != nil {
+			glog.V(logger.Info).Infof("could not get extraBalance for DAO: %v %v", i, err)
+			return nil, err
+		}
+		hfData.ExtraBalances = append(hfData.ExtraBalances, bal)
+	}
+
+	for i := 0; i < len(daos); i++ {
+		hfData.ChildDAOs = append(hfData.ChildDAOs, daos[i].Address())
+	}
+
+	return hfData, nil
+}
+
+func traverseDAOs(d *dao.DAO, eth *Ethereum) ([]*dao.DAO, error) {
```

Comment:
Thanks for the explanation. I strongly agree that evm protocol changes should be minimized/avoided.

I like that balances are gathered up from the child DAOs is exposed as a network parameter and confirmed across clients. 

What would happen if a transaction that manipulated DAO state, was introduced at the same time as the fork, and one miner included the transaction in their fork-block and another did not include it. Presumably that will just be resolved as an ordinary/temporary fork by miners?